spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java

```diff
@@ -864,18 +861,89 @@ else if (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic)
 					"will be created before the request Message is sent.");
 		}
 		MessageProducer messageProducer = null;
-		MessageConsumer messageConsumer = null;
 		try {
 			messageProducer = session.createProducer(requestDestination);
 			this.sendRequestMessage(jmsRequest, messageProducer, priority);
 			String messageId = jmsRequest.getJMSMessageID().replaceAll("'", "''");
 			String messageSelector = "JMSCorrelationID = '" + messageId + "'";
-			messageConsumer = session.createConsumer(replyTo, messageSelector);
-			return this.receiveReplyMessage(messageConsumer);
+			return retryableReceiveReply(session, replyTo, messageSelector);
 		}
 		finally {
 			JmsUtils.closeMessageProducer(messageProducer);
-			JmsUtils.closeMessageConsumer(messageConsumer);
+		}
+	}
+
+
+	/*
+	 * If the replyTo is not temporary, and the connection is lost while waiting for a reply, reconnect for
+	 * up to receiveTimeout.
+	 */
+	private javax.jms.Message retryableReceiveReply(Session session, Destination replyTo, String messageSelector)
+			throws JMSException {
+		Connection consumerConnection = null;
+		Session consumerSession = session;
+		MessageConsumer messageConsumer = null;
+		long now = System.currentTimeMillis();
+		JMSException exception = null;
+		boolean isTemporaryReplyTo = replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic;
+		try {
+			do {
+				try {
+					messageConsumer = consumerSession.createConsumer(replyTo, messageSelector);
+					javax.jms.Message reply = receiveReplyMessage(messageConsumer);
+					if (reply == null) {
+						if (this.receiveTimeout < 0 || now + this.receiveTimeout > System.currentTimeMillis()) {
```

Comment:
Minor, of course, but it isn't `now` here already after several iterations.
How about `receiveStart` ?